### ISSUE TITLE
PSR12.Files.FileHeader misidentifies file header in mixed content file

### ISSUE THREAD

**Issue by jrfnl** [squizlabs/PHP_CodeSniffer#2732]
For the following code sample I would expect an error on line 1 about the missing blank line between the PHP open tag and the file docblock.

```php
<?php
/**
 * Template file docblock.
 *
 * @package Vendor\Package
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

?>

<p><?php echo 'Some text string'; ?></p>

<?php

/**
 * Docblock.
 */
include 'foo.php';

```

The above mentioned expected error is not being thrown.

Instead I receive the following errors:
```
-------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
-------------------------------------------------------------------------
 16 | ERROR | [ ] The file header must be the first content in the file
 20 | ERROR | [x] Header blocks must be separated by a single blank line
-------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-------------------------------------------------------------------------
```

**Comment by jrfnl** [comment]
Found another example of the same issue:
```php
<?php
/**
 * File docblock.
 *
 * @package Vendor\Package
 */

class Foo {
	/**
	 * Function docblock.
	 */
	public function bar() {
		do_something();
		?>
		<p>Demo</p>

		<?php
		/**
		 * Arbitrary docblock.
		 */
		api_call();
	}
}
```

**Comment by gsherwood** [comment]
I had to change a fair bit to get this working, mostly because it's very hard to determine if a comment is a file header or not. But now the sniff processes the entire file looking for the first thing that looks like a file header, and reverts to the very first PHP block if it can't find one. This lets the "newline after open tag" rule apply in all cases.

It also stops looking if it finds an OO token so it wont pick up things that look like file headers but are clearly not. This fixed the issue from the above comment.

Thanks for providing those code samples.

**Comment by jrfnl** [comment]
@gsherwood Thanks for fixing this! I appreciate the effort which has gone into it.
